packages/cel/src/type.ts

```diff
@@ -12,117 +12,310 @@
-  ? CelType :
-  T extends CelScalar.ANY
-  ? CelVal :
-  T extends CelScalar.INT
-  ? bigint :
-  T extends CelScalar.UINT
-  ? CelUint :
-  T extends CelScalar.DOUBLE
-  ? number :
-  T extends CelScalar.BOOL
-  ? boolean :
-  T extends CelScalar.STRING
-  ? string :
-  T extends CelScalar.BYTES
-  ? Uint8Array :
-  T extends CelScalar.NULL
-  ? null :
-  T extends CelListValueType
-  ? CelList :
-  T extends CelMapValueType<infer K, infer V>
-  ? CelMap :
-  T extends DescMessage
-  ? MessageShape<T> :
-  T;
+export type CelOutput<T extends CelType = CelType> = 
+  T extends typeof CelScalar.DYN ? celOutput : celOutput<T>; // Avoids the infinite recursion.
 
 // biome-ignore format: Ternaries
-export type TupleTypeOf<T extends readonly CelValueType[]> =
+type celOutput<T extends CelType = CelType> = 
+    T extends typeof CelScalar.NULL     ? null
+  : T extends CelObjectType<infer Desc> ? MessageShape<Desc>
+  : celValue<T>;
+
+// biome-ignore format: Ternaries
+export type CelOutputTuple<T extends readonly CelType[]> =
   T extends readonly [
-    infer First extends CelValueType,
-    ...infer Rest extends CelValueType[],
+    infer First extends CelType,
+    ...infer Rest extends CelType[],
   ]
-    ? [TypeOf<First>, ...TupleTypeOf<Rest>]
-    // biome-ignore lint/suspicious/noExplicitAny: This is only valid in the case of TupleTypeOf<CelValueType[]>     
-    : CelValueType[] extends T ? any[] : [];
+    ? [CelOutput<First>, ...CelOutputTuple<Rest>]
+    // biome-ignore lint/suspicious/noExplicitAny: This is only valid in the case of CelTupleValue<CelValueType[]>     
+    : CelType[] extends T ? any[] : [];
+
+/**
+ * Get the CELType of a CelValue.
+ */
+export function celType(v: CelValue): CelType {
+  switch (typeof v) {
+    case "bigint":
+      return CelScalar.INT;
+    case "boolean":
+      return CelScalar.BOOL;
+    case "number":
+      return CelScalar.DOUBLE;
+    case "string":
+      return CelScalar.STRING;
+    case "object":
+      switch (true) {
+        case v === null:
+        case isNullMessage(v):
+          return CelScalar.NULL;
+        case v instanceof Uint8Array:
+          return CelScalar.BYTES;
+        case isReflectMessage(v):
+          return objectType(v.desc);
+        case isCelList(v):
+          return listType(CelScalar.DYN);
+        case isCelMap(v):
+          return mapType(CelScalar.DYN, CelScalar.DYN);
+        case isCelUint(v):
+          return CelScalar.UINT;
+        default:
+          // This can also be a case statement, but TS fails to
+          // narrow the type.
+          if (isObjectCelType(v)) {
+            return typeType(v);
+          }
+      }
+  }
+  throw new Error(`Not a CEL value: ${v}`);
+}
+
+/**
+ * Returns true if the given value is a CelType.
+ */
+export function isCelType(v: unknown): v is CelType {
+  return typeof v === "object" && v !== null && isObjectCelType(v);
+}
+
+export function isObjectCelType(v: NonNullable<object>): v is CelType {
```

Comment:
This could be a pattern for other types as well which will skip the object and not null check.